src/EditorFeatures/CSharpTest/Completion/CompletionProviders/DeclarationNameCompletionProviderTests.cs

```diff
@@ -0,0 +1,462 @@
+            try
+            {
+                var newOptions = oldOptions.WithChangedOption(new OptionKey(SimplificationOptions.NamingPreferences, LanguageNames.CSharp), AsyncMethodsEndWithAsync());
+                workspace.Options = newOptions;
+                await VerifyItemExistsAsync(markup, "GetCAsync");
+            }
+            finally
+            {
+                workspace.Options = oldOptions;
+            }
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async void MethodDeclaration1()
+        {
+            var markup = @"
+public class C
+{
+    virtual C $$
+}
+";
+            await VerifyItemExistsAsync(markup, "GetC");
+            await VerifyItemIsAbsentAsync(markup, "C");
+            await VerifyItemIsAbsentAsync(markup, "c");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async void WordBreaking1()
+        {
+            var markup = @"
+using System.Threading;
+public class C
+{
+    CancellationToken $$
+}
+";
+            await VerifyItemExistsAsync(markup, "cancellationToken");
+            await VerifyItemExistsAsync(markup, "cancellation");
+            await VerifyItemExistsAsync(markup, "token");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async void WordBreaking2()
+        {
+            var markup = @"
+interface I {}
+public class C
+{
+    I $$
+}
+";
+            await VerifyItemExistsAsync(markup, "I");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async void WordBreaking3()
+        {
+            var markup = @"
+interface II {}
+public class C
+{
+    I $$
+}
+";
+            await VerifyItemExistsAsync(markup, "I");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async void WordBreaking4()
+        {
+            var markup = @"
+interface IFoo {}
+public class C
+{
+    IFoo $$
+}
+";
+            await VerifyItemExistsAsync(markup, "Foo");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async void WordBreaking5()
+        {
+            var markup = @"
+class AWonderfullyLongClassName {}
+public class C
+{
+    AWonderfullyLongClassName $$
+}
+";
+            await VerifyItemExistsAsync(markup, "A");
+            await VerifyItemExistsAsync(markup, "AWonderfully");
+            await VerifyItemExistsAsync(markup, "AWonderfullyLong");
+            await VerifyItemExistsAsync(markup, "AWonderfullyLongClass");
+            await VerifyItemExistsAsync(markup, "Name");
+            await VerifyItemExistsAsync(markup, "ClassName");
+            await VerifyItemExistsAsync(markup, "LongClassName");
+            await VerifyItemExistsAsync(markup, "WonderfullyLongClassName");
+            await VerifyItemExistsAsync(markup, "AWonderfullyLongClassName");
+        }
```

Comment:
This test harness can't test ordering, so I added a test in CSharpCompletionCommandHandlerTests that tests the ordering. Right now it orders long->short with LTR/RTL interleaved